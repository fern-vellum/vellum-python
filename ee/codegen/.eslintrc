{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "extends": [
    "eslint:recommended",
    "plugin:import/recommended",
    "plugin:import/typescript",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:prettier/recommended"
  ],
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "node": {
        "extensions": [".js", ".jsx", ".ts", ".tsx"]
      },
      "typescript": {
        "project": "**/tsconfig.json"
      }
    }
  },
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module",
    "ecmaFeatures": {
      "experimentalObjectRestSpread": true
    }
  },
  "env": {
    "es6": true,
    "node": true
  },
  "rules": {
    "import/order": [
      "error",
      {
        "alphabetize": {
          "order": "asc",
          "orderImportKind": "asc",
          "caseInsensitive": false
        },
        "groups": [
          "builtin",
          "external",
          "parent",
          "sibling",
          "index",
          "internal",
          "object",
          "type"
        ],
        "newlines-between": "always"
      }
    ],
    "no-console": ["error", { "allow": ["warn", "error"] }],
    "@typescript-eslint/no-explicit-any": "error",
    "@typescript-eslint/no-non-null-assertion": "error",
    "@typescript-eslint/no-unused-vars": [
      "error",
      {
        "argsIgnorePattern": "^_[_\\d]*$",
        "varsIgnorePattern": "^_[_\\d]*$"
      }
    ],
    "@typescript-eslint/ban-ts-comment": 0,
    "@typescript-eslint/no-namespace": "off",
    "no-relative-import-paths/no-relative-import-paths": [
      "error",
      {
        "allowSameFolder": true
      }
    ],
    "import/export": "off",
    "unused-imports/no-unused-imports": "error"
  },
  "plugins": [
    "@typescript-eslint",
    "import",
    "no-relative-import-paths",
    "unused-imports"
  ],
  "ignorePatterns": ["lib", "scripts"]
}
