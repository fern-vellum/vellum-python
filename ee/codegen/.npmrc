@vellum-ai:registry=https://us-central1-npm.pkg.dev/vocify-prod/vellum-private-npm/
//us-central1-npm.pkg.dev/vocify-prod/vellum-private-npm/:always-auth=true
