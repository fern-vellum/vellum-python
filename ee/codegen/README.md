# Workflow Codegen

This is used to generate the code representation of a Vellum Workflow given its JSON display representation.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Updating this README](#updating-this-readme)
- [Setup](#setup)
- [Testing](#testing)
  - [Tips for Writing Tests](#tips-for-writing-tests)
    - [Unit Testing](#unit-testing)
    - [Integration Testing](#integration-testing)
- [Developing Alongside Fern](#developing-alongside-fern)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Updating this README

This README is generated using [doctoc](https://github.com/thlorenz/doctoc)

To install doctoc, run `npm install -g doctoc`

To update the table of contents, run `doctoc README.md`

## Setup

1. Install `nvm` to manage Node versions by following the instructions [here](https://github.com/nvm-sh/nvm) or
   check your `nvm` version with `nvm ---version`.
2. Install the version of Node we rely on by running:
   ```bash
   nvm install v18.12.1
   ```
3. Ensure you're using Node 18 by default via:
   ```bash
   nvm alias default v18.12.1
   ```
4. Use `npm` to install the project dependencies:
   ```bash
   npm install
   ```
5. If you haven't already, go to the CONTRIBUTING.md guide and run the setup steps there

## Testing

To run unit tests:

For these tests to run successfully, you'll need to run the poetry virtual environment.
The steps to set this up are in the README at the root of this repo.

```bash
npm run test
```

Many tests rely on snapshots. If you've changed logic such that the snapshot is intentionally different, you
can update snapshots by running:

```bash
npm run test:update
```

To validate types:

```bash
npm run types
```

### Tips for Writing Tests

There are two types of tests in this package.

#### Unit Tests

Unit tests test individual generators in isolation. For example, you might write a unit test to validate that
`TextSearchNode` generates the correct python code given its JSON representation.

Unit tests should use snapshot testing to validate the generated code. For example,

```typescript
it("TextSearchNode", async () => {
  const node = new TextSearchNode({
    workflowContext,
    nodeData,
  });

  node.getNodeFile().write(writer);
  expect(await writer.toStringFormatted()).toMatchSnapshot();
});
```

#### Integration Tests

Integration tests test the entire code generation process for a complete Workflow. The way they work is we automatically
traverse the `workflows/codegen/tests/fixtures` directory and run the tests on each `.json` file in that directory.

`codegen/src/__test__/project.test.ts` is responsible for running these integration tests.

## Developing Alongside Fern

We use fern's python [codegen package](https://github.com/fern-api/fern/tree/main/generators/python-v2/codegen) to
generate the python code representation of a Vellum Workflow. In some cases, you might find that fern doesn't yet have
support for constructing the python syntax you need. In these cases, we can contribute to fern's open source repo directly.

It's usually best to implement your changes in fern locally and confirm that they fulfill your needs in this repo
before opening the PR to fern with your changes. To make changes to fern's python codegen package and test those
changes in the workflow codegen package locally, follow these steps:

1. Clone the fern repository
   ```bash
   git clone git@github.com:fern-api/fern.git
   ```
2. Update the `"@fern-api/python-ast"` dependency in `package.json` to point to the local fern repository.
   It should be the relative file path to the `ast` directory in this repository.
   ```json
   "@fern-api/python-ast": "file:../../../fern-api/fern/generators/python-v2/ast/lib",
   ```
3. Run `npm install` to install the updated dependency.
4. Navigate to the `fern` repo and be sure to pull the `vargas/publish-python-ast` branch.
5. Make changes to the fern python ast package.
6. After making changes, run `pnpm dist` within `fern/generators/python-v2/ast` to compile the changes.
7. Navigate back to this repo and make use of these changes
8. Open a PR of the commit in its own branch (separate from the `vargas/publish-python-ast` branch) in the fern repository to merge your changes.
9. Work with a Vellum Admin to cut a new release of the package as we'll need to rebase the `vargas/publish-python-ast`
   branch
10. Update the dependency in this repo to the new version.

## Publishing

This package is meant to be versioned along with the Vellum Python SDK. When incrementing to match the latest Vellum Python SDK version, run:

```bash
npm version patch|minor|major
```

However, there are times where we need to publish a new version of this package without incrementing the Vellum Python SDK version. In this case, we increment the version using the prerelease strategy:

```bash
npm run version:codegen
```

This will bump the version from `0.2.6` to `0.2.6-post1`, `0.2.6-post2`, etc. It's important to note that **this is not SemVer Compliant**. SemVer doesn't have a concept of doing postreleases well, and technically speaking, does not consider `0.2.6-post1` to be later than `0.2.6`. Given that this package is only used internally for now, this is not a huge deal, but it's something to be aware of. If we wanted to be SemVer compliant, then we would have to choose a different versioning strategy. The alternatives are not ideal, but they are:

- Version codegen and the sdk independently.
- Version the SDK with just majors and minors going forward, opening up patch versions for codegen.

We have a GitHub Action that will publish all of the patch versions to NPM (i.e, the ones without a `-post` suffix). If that action fails, or we need to publish a post-release version, we'll need to publish manually.

To publish a package manually, you'll need to run the following commands locally:

```bash
npm run gar-login
npm publish
```

_Note: For some reason, the `gar-login` command doesn't work as part of the `prepublish` script, so it needs to be run separately._
