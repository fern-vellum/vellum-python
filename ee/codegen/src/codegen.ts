import {
  ErrorLogFile,
  InitFile,
  Inputs,
  NodeInput,
  VellumValue,
  VellumVariable,
  Workflow,
} from "./generators";

export function vellumVariable(args: VellumVariable.Args): VellumVariable {
  return new VellumVariable(args);
}

export function vellumValue(args: VellumValue.Args): VellumValue {
  return new VellumValue(args);
}

export function nodeInput(args: NodeInput.Args): NodeInput {
  return new NodeInput(args);
}

export function inputs(args: Inputs.Args): Inputs {
  return new Inputs(args);
}

export function workflow(args: Workflow.Args): Workflow {
  return new Workflow(args);
}

export function initFile(args: InitFile.Args): InitFile {
  return new InitFile(args);
}

export function errorLogFile(args: ErrorLogFile.Args): ErrorLogFile {
  return new ErrorLogFile(args);
}
