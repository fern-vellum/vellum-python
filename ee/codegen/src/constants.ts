/* Module paths */
export const VELLUM_CLIENT_MODULE_PATH = ["vellum"] as const;

/* Class names */
export const OUTPUTS_CLASS_NAME = "Outputs";
export const PORTS_CLASS_NAME = "Ports";
export const DEFAULT_PORT_NAME = "default";

/* File names */
export const INIT_FILE_NAME = "__init__.py";

export const GENERATED_WORKFLOW_MODULE_NAME = "workflow";

export const GENERATED_INPUTS_MODULE_NAME = "inputs";

export const GENERATED_NODES_MODULE_NAME = "nodes";

export const GENERATED_DISPLAY_MODULE_NAME = "display";
export const GENERATED_NODES_PATH = [GENERATED_NODES_MODULE_NAME] as const;

export const GENERATED_DISPLAY_NODE_MODULE_PATH = [
  GENERATED_DISPLAY_MODULE_NAME,
  GENERATED_NODES_MODULE_NAME,
] as const;

export const GENERATED_NESTED_NODE_MODULE_NAME = "node";
