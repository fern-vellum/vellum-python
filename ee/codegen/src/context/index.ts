export { createNodeContext } from "./node-context";
export { WorkflowContext } from "./workflow-context";
