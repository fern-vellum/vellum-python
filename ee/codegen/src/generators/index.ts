export { ErrorLogFile } from "./error-log-file";
export { InitFile } from "./init-file";
export { Inputs } from "./inputs";
export { NodeInput } from "./node-inputs";
export { Workflow } from "./workflow";
export { VellumVariable } from "./vellum-variable";
export { VellumValue } from "./vellum-variable-value";
