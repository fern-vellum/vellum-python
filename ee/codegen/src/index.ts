export * as codegen from "./codegen";
export * as errors from "src/generators/errors";
export * as project from "./project";
