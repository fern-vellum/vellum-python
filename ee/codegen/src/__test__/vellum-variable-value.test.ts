import { Writer } from "@fern-api/python-ast/core/Writer";

import * as codegen from "src/codegen";

describe("VellumValue", () => {
  let writer: Writer;

  beforeEach(() => {
    writer = new Writer();
  });

  describe("STRING", () => {
    it("should write a STRING value correctly", async () => {
      const stringValue = codegen.vellumValue({
        vellumValue: {
          type: "STRING",
          value: "Hello, World!",
        },
      });
      stringValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("NUMBER", () => {
    it("should write a NUMBER value correctly", async () => {
      const numberValue = codegen.vellumValue({
        vellumValue: {
          type: "NUMBER",
          value: 42,
        },
      });
      numberValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("CHAT_HISTORY", () => {
    it("should write a CHAT_HISTORY value with just text", async () => {
      const chatHistoryValue = codegen.vellumValue({
        vellumValue: {
          type: "CHAT_HISTORY",
          value: [
            {
              role: "USER",
              text: "Hello, AI!",
            },
          ],
        },
      });
      chatHistoryValue.write(writer);
      expect(await writer.toString()).toMatchSnapshot();
    });

    it("should write a CHAT_HISTORY value with a string content", async () => {
      const chatHistoryValue = codegen.vellumValue({
        vellumValue: {
          type: "CHAT_HISTORY",
          value: [
            {
              role: "USER",
              content: { type: "STRING", value: "Hello, AI!" },
            },
          ],
        },
      });
      chatHistoryValue.write(writer);
      expect(await writer.toString()).toMatchSnapshot();
    });
  });

  describe.skip("SEARCH_RESULTS", () => {
    it("should write a SEARCH_RESULTS value correctly", async () => {
      const searchResultsValue = codegen.vellumValue({
        vellumValue: {
          type: "SEARCH_RESULTS",
          value: [],
        },
      });
      searchResultsValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("JSON", () => {
    it("should write a JSON value correctly", async () => {
      const jsonValue = codegen.vellumValue({
        vellumValue: {
          type: "JSON",
          value: {
            key: "value",
            nested: { array: [1, 2, 3] },
          },
        },
      });
      jsonValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("ERROR", () => {
    it("should write a ERROR value correctly", async () => {
      const errorValue = codegen.vellumValue({
        vellumValue: {
          type: "ERROR",
          value: {
            message: "This is an error!",
            code: "INTERNAL_SERVER_ERROR",
          },
        },
      });
      errorValue.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
