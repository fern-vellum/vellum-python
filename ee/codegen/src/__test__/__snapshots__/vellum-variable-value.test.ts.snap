// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VellumValue > AUDIO > should write a AUDIO value correctly 1`] = `
"VellumAudio(src="https://example.com/audio.mp3")
"
`;

exports[`VellumValue > CHAT_HISTORY > should write a CHAT_HISTORY value with a string content 1`] = `"[ChatMessageRequest(role="USER", content=StringChatMessageContentRequest(value="Hello, AI!"))]"`;

exports[`VellumValue > CHAT_HISTORY > should write a CHAT_HISTORY value with just text 1`] = `"[ChatMessageRequest(role="USER", text="Hello, AI!")]"`;

exports[`VellumValue > ERROR > should write a ERROR value correctly 1`] = `
"VellumError(message="This is an error!", code="INTERNAL_SERVER_ERROR")
"
`;

exports[`VellumValue > FUNCTION_CALL > should write a FUNCTION_CALL value correctly 1`] = `
"FunctionCall(arguments={"key": "value"}, name="test", id="123")
"
`;

exports[`VellumValue > IMAGE > should write a IMAGE value correctly 1`] = `
"VellumImage(src="https://example.com/image.png")
"
`;

exports[`VellumValue > JSON > should write a JSON value correctly 1`] = `
"{"key": "value", "nested": {"array": [1, 2, 3]}}
"
`;

exports[`VellumValue > NUMBER > should write a NUMBER value correctly 1`] = `
"42
"
`;

exports[`VellumValue > SEARCH_RESULTS > should write a SEARCH_RESULTS value correctly 1`] = `
"[
    SearchResult(
        text="Hello, World!", score=0.5, keywords=["hello", "world"], document=document
    ),
    SearchResult(
        text="Hello, AI!", score=0.7, keywords=["hello", "ai"], document=document
    ),
]
"
`;

exports[`VellumValue > STRING > should write a STRING value correctly 1`] = `
""Hello, World!"
"
`;
